Server/Components/CAPI/Impl/ComponentManager.hpp

```diff
@@ -0,0 +1,256 @@
+				{
+				case EventReturnHandler::StopAtFalse:
+					if (!ret)
+					{
+						delete[] eventArgs.list;
+						return false;
+					}
+					break;
+				case EventReturnHandler::StopAtTrue:
+					if (ret)
+					{
+						delete[] eventArgs.list;
+						return true;
+					}
+					break;
+				case EventReturnHandler::None:
+				default:
+					break;
+				}
+				result = ret;
+			}
+		}
+
+		delete[] eventArgs.list;
+		return result;
+	}
+};
+
+/// Get player from players pool
+template <typename PlayerDataType>
+inline PlayerDataType* GetPlayerData(IPlayer* player)
+{
+	if (player == nullptr)
+	{
+		return nullptr;
+	}
+	return queryExtension<PlayerDataType>(*player);
+}
+
+/// Cast object pointer to entity type, return fail_ret if not available
+#define ENTITY_CAST_RET(entity_type, entity, output, failret) \
+	if (!entity)                                              \
+		return failret;                                       \
+	entity_type* output = reinterpret_cast<entity_type*>(entity)
+
+#define ENTITY_CAST(entity_type, entity, output) \
+	if (!entity)                                 \
+		return;                                  \
+	entity_type* output = reinterpret_cast<entity_type*>(entity)
+
+#define POOL_ENTITY_RET(pool, entity_type, entity, output, failret) \
+	if (!ComponentManager::Get()->pool)                             \
+		return failret;                                             \
+	auto pool = ComponentManager::Get()->pool;                      \
+	ENTITY_CAST_RET(entity_type, entity, output, failret)
+
+#define POOL_ENTITY(pool, entity_type, entity, output) \
+	if (!ComponentManager::Get()->pool)                \
+		return;                                        \
+	auto pool = ComponentManager::Get()->pool;         \
+	ENTITY_CAST(entity_type, entity, output)
+
+#define PLAYER_POOL_ENTITY_RET(player, pool_type, entity_type, entity, entity_output, failret) \
+	auto playerData = GetPlayerData<pool_type>(player);                                        \
+	if (playerData == nullptr)                                                                 \
+		return failret;                                                                        \
+	entity_type* entity_output = reinterpret_cast<entity_type*>(entity);                       \
+	if (entity_output == nullptr)                                                              \
+		return failret;
+
+#define PLAYER_DATA_RET(player, entity_type, entity_output, failret) \
+	auto entity_output = GetPlayerData<entity_type>(player);         \
+	if (entity_output == nullptr)                                    \
+	return failret
+
+#define COPY_STRING_TO_CAPI_STRING_VIEW(output, src, len_) \
+	if (output)                                            \
+	{                                                      \
+		output->len = len_;                                \
+		memcpy(output->data, src, uint32_t(len_));         \
+	}
+
+#define SET_CAPI_STRING_VIEW(output, str_view)             \
```

Comment:
The lifetime of SET_CAPI_STRING_VIEW & CREATE_CAPI_STRING_VIEW vs COPY_STRING_TO_CAPI_STRING_VIEW can cause some issues. I think it should always be copied into or, if you don't want to copy because of performance, the copyable one should be separated from the string view into a separate struct, and there should be documentation about the differences between them. A user currently won't know when to copy string results for permanent storage.

Reply:
done